Updates to enable soil DA

ClaraDraper-NOAA opened the pull request:
# Description
Updates to enable soil DA: 
-update the orog_ver, to include the fix files needed for the soil DA
-removed the temporary hard-coded directory with the fix files
-moved the regridding code out of GDASApp and UFS_UTILS  
-added a soil DA CI test

  Resolves [#3390](https://github.com/NOAA-EMC/global-workflow/issues/3390)
  Resolves [#3392](https://github.com/NOAA-EMC/global-workflow/issues/3392) 
  Refs https://github.com/JCSDA/spack-stack/issues/1562#issuecomment-2725573898
  Resolves https://github.com/NOAA-EMC/global-workflow/issues/3484
 
# Type of change
- [ x] Bug fix (fixes something broken)
- [ ] New feature (adds functionality)
- [ ] Maintenance (code refactor, clean-up, new CI test, etc.)

# Change characteristics
- Is this a breaking change (a change in existing functionality)? NO
- Does this change require a documentation update? NO
- Does this change require an update to any of the following submodules? YES 

  - [ ] [UFS-utils](https://github.com/ufs-community/UFS_UTILS/pull/1038)  
  - [ ] [Spack-stack ](https://github.com/JCSDA/spack-stack/issues/1562#issuecomment-2725573898)

For completeness, should also merge: https://github.com/NOAA-EMC/DA-utils/pull/24
(this removed the the code from GDASApp, but is not strictly required for the current PR) 

# How has this been tested?

Cloned, built on hera. Ran C96C48_hybatmDA.yaml and C96C48_hybatmsoilDA.yaml tests.

-->

# Checklist
- [ ] Any dependent changes have been merged and published
- [x ] My code follows the style guidelines of this project
- [x ] I have performed a self-review of my own code
- [x ] I have commented my code, particularly in hard-to-understand areas
- [x ] I have documented my code, including function, input, and output descriptions
- [x ] My changes generate no new warnings
- [ x] New and existing tests pass with my changes
- [ x] This change is covered by an existing CI test or a new one has been added
- [ x] Any new scripts have been added to the .github/CODEOWNERS file with owners
- [ x] I have made corresponding changes to the system documentation if necessary

ClaraDraper-NOAA committed:
Updates to
1. move regrid code from DA-utils to UFS_UTILS
2. update the ufs fix version
3. add a soil DA test.

ClaraDraper-NOAA committed:
Merge branch 'develop' into feature/mv_regridding

ClaraDraper-NOAA committed:
Tidy up.

ClaraDraper-NOAA commented:
Created as a draft, while we wait for the dependent PRs to be merged.

ClaraDraper-NOAA committed:
Updated spack-stack version so can load esmf/8.6.0

JessicaMeixner-NOAA commented on ci/cases/pr/C96C48_hybatmsoilDA.yaml:
```diff
@@ -0,0 +1,24 @@
+  start: cold
+  yaml: {{ HOMEgfs }}/ci/cases/yamls/atmsoilDA_defaults_ci.yaml
+
+skip_ci_on_hosts:
```
@ClaraDraper-NOAA are these tests skipped to save time or because things do not work on these other machines? Particularly concerned about orion, hercules and gaeac6 which are being considered for GFSv17 retrospectives.

ClaraDraper-NOAA replied:
To save time. I've turned the tests on for all machines.

JessicaMeixner-NOAA replied:
Thanks for the explanation @ClaraDraper-NOAA   and confirming that it should work everywhere.

ClaraDraper-NOAA committed:
run soil DA test on all machines.

CatherineThomas-NOAA commented:
Does the sorc/link_workflow.sh script also need to be updated to point to the new repo for the regrid executable?

ClaraDraper-NOAA commented:
> Does the sorc/link_workflow.sh script also need to be updated to point to the new repo for the regrid executable?

At the moment the regrid executable is being [accessed](https://github.com/ClaraDraper-NOAA/global-workflow/blob/af974073fce7afcd22840753305e058d3067ec68/parm/config/gfs/config.esfc#L34) from the ufs_utils/exec directory. Would it be better to link it with link_worlflow.sh to the global_workflow/exec and access it from there?

CatherineThomas-NOAA commented:
> At the moment the regrid executable is being [accessed](https://github.com/ClaraDraper-NOAA/global-workflow/blob/af974073fce7afcd22840753305e058d3067ec68/parm/config/gfs/config.esfc#L34) from the ufs_utils/exec directory. Would it be better to link it with link_worlflow.sh to the global_workflow/exec and access it from there?

My guess would be that, yes, it would be better to link it, but maybe @aerorahul or @DavidHuber-NOAA can chime in?

ClaraDraper-NOAA commented:
> > At the moment the regrid executable is being [accessed](https://github.com/ClaraDraper-NOAA/global-workflow/blob/af974073fce7afcd22840753305e058d3067ec68/parm/config/gfs/config.esfc#L34) from the ufs_utils/exec directory. Would it be better to link it with link_worlflow.sh to the global_workflow/exec and access it from there?
> 
> My guess would be that, yes, it would be better to link it, but maybe @aerorahul or @DavidHuber-NOAA can chime in?

I'll just move it. That's more consistent with what is done elsewhere.

ClaraDraper-NOAA committed:
Linking sfc_regrid executable into global_workflow exec.

ClaraDraper-NOAA commented:
> > > At the moment the regrid executable is being [accessed](https://github.com/ClaraDraper-NOAA/global-workflow/blob/af974073fce7afcd22840753305e058d3067ec68/parm/config/gfs/config.esfc#L34) from the ufs_utils/exec directory. Would it be better to link it with link_worlflow.sh to the global_workflow/exec and access it from there?
> > 
> > 
> > My guess would be that, yes, it would be better to link it, but maybe @aerorahul or @DavidHuber-NOAA can chime in?
> 
> I'll just move it. That's more consistent with what is done elsewhere.

done.

ClaraDraper-NOAA committed:
Add new spack-stack to hercules, gaea[56], and added explicit
versions for the new modules on other machines.
Skipping soilDA test on orion, while waiting for spack-stack issue
to be resolved.

ClaraDraper-NOAA committed:
Amend previous commit.

ClaraDraper-NOAA committed:
Merge branch 'develop' into feature/mv_regridding

ClaraDraper-NOAA committed:
Bugfix in GAEA env files.

ClaraDraper-NOAA committed:
bugfix to Gaea[56] env files.

ClaraDraper-NOAA committed:
remove errornous dependency of gfs_sfcanl on eupd

ClaraDraper-NOAA committed:
py norms.

ClaraDraper-NOAA committed:
Merge branch 'develop' into feature/mv_regridding

ClaraDraper-NOAA committed:
Updated modules on orion.

ClaraDraper-NOAA committed:
script change to allow monitoring of 2m obs.

ClaraDraper-NOAA committed:
Prevent regridding exec from being over-written.

ClaraDraper-NOAA committed:
Merge branch 'develop' into feature/mv_regridding

ClaraDraper-NOAA commented:
Update: The UFS_UTILS PR that has been holding this up is close to being merged. Given it's been a while, I'm cloning a clean version and re-doing the tests. I'll update again when everything is ready.

ClaraDraper-NOAA committed:
Restricting regid_ntasks to 6, for reproducibility.

ClaraDraper-NOAA committed:
Merge branch 'develop' into feature/mv_regridding

ClaraDraper-NOAA committed:
Updated UFS_UTILS hash.

ClaraDraper-NOAA commented:
The dependent PRs are all through - I think this is ready to go. 

I'm re-running the tests for the latest version, and will confirm once they're through.

DavidHuber-NOAA commented on modulefiles/module_base.gaeac5.lua:
```diff
@@ -15,6 +15,7 @@ load(pathJoin("cdo", (os.getenv("cdo_ver") or "None")))
 load(pathJoin("hdf5", (os.getenv("hdf5_ver") or "None")))
 load(pathJoin("netcdf-c", (os.getenv("netcdf_c_ver") or "None")))
 load(pathJoin("netcdf-fortran", (os.getenv("netcdf_fortran_ver") or "None")))
+load(pathJoin("esmf", (os.getenv("esmf_ver") or "None")))
```
Does ESMF actually need to be loaded at runtime?  Usually, this library is statically linked.

ClaraDraper-NOAA replied:
I think I tried it without, and it didn't work. Let me double-check that.

ClaraDraper-NOAA replied:
Without that line, it fails finding the pio library, which is needed by ESMF (or I should say, I've always assumed it's needed by ESMF, as that's always the first failure when ESMF isn't loaded).

DavidHuber-NOAA replied:
OK, then I'm fine with keeping it.  Resolving conversation.

ClaraDraper-NOAA commented:
> The dependent PRs are all through - I think this is ready to go.
> 
> I'm re-running the tests for the latest version, and will confirm once they're through.

The 3DVar atmos and soilDA tests passed with the latest develop merged in.

KateFriedman-NOAA commented on versions/build.gaeac6.ver:
```diff
@@ -1,5 +1,5 @@
 export stack_intel_ver=2023.2.0
 export stack_cray_mpich_ver=8.1.29
-export spack_env=gsi-addon
+export spack_env=gsi-addon-dev-fms-2024.01
```
This will impact the entire system. What is the different between `gsi-addon` and `gsi-addon-dev-fms-2024.01`? Just a different fms module available?

Tagging @aerorahul and @DavidHuber-NOAA for their inputs/knowledge.

ClaraDraper-NOAA replied:
@DavidHuber-NOAA was involved in this, I'll let him respond. 

Various people have run some of the tests with this upgrade on hera, orion, and C6.

DavidHuber-NOAA replied:
The `gsi-addon-dev-fms-2024.01` environment is an updated spack-stack 1.6.0 environment.  It is identical to `gsi-addon` except for 3 libraries that I know of: fms, g2, and esmf are all upgraded to newer versions.  This is a chained environment, meaning that it is built off of a parent.  The parent in this case is `fms-2024.01`, which is the environment that the UFS and UFS_Utils are currently using (the `gsi-addon-dev-fms-2024.01` just replaces the `bufr`, `met`, and `metplus` versions as was the case for the `gsi-addon` environment).  On a side note, it may be possible with this upgrade to remove calls to `load_ufswm_modules.sh`, but that would need to be tested and isn't something that should go into this PR.

KateFriedman-NOAA replied:
Got it, thanks for the history on this new env @DavidHuber-NOAA ! 

> On a side note, it may be possible with this upgrade to remove calls to load_ufswm_modules.sh, but that would need to be tested and isn't something that should go into this PR.

Agreed, that would be great!

KateFriedman-NOAA commented on scripts/exgdas_enkf_update.sh:
```diff
@@ -204,7 +204,7 @@ for imem in $(seq 1 $NMEM_ENS); do
       ${NLN} "${COMIN_ATMOS_HISTORY_MEM_PREV}/${GPREFIX}atmf00${FHR}${ENKF_SUFFIX}.nc" \
          "sfg_${PDY}${cyc}_fhr0${FHR}_${memchar}"
-      if [[ "${DO_GSISOILDA}" = "YES" ]]; then
+      if [[ "${hofx_2m_sfcfile}" = ".true." ]]; then
```
Where does the value of `hofx_2m_sfcfile` get set?

ClaraDraper-NOAA replied:
It's set in [config.anal](https://github.com/NOAA-EMC/global-workflow/blob/16219a3e6904662b774b3b23edc576b885737b3b/parm/config/gfs/config.anal#L52)

I made this change so that we can monitor the 2m obs without assimilating them.

CoryMartin-NOAA replied:
Is it defined at the top of this script in some style like:
`hofx_2m_sfcfile=${hofx_2m_sfcfile:-".false."}`
if not, we should probably do that

KateFriedman-NOAA replied:
Thanks for pointing me to where it's set @ClaraDraper-NOAA . I also second @CoryMartin-NOAA's suggestion.

ClaraDraper-NOAA replied:
done.

CoryMartin-NOAA commented on ci/cases/pr/C96C48_hybatmsoilDA.yaml:
Can we just combine this with the existing CI test?

KateFriedman-NOAA replied:
I second this. Could this be combined with the existing `ci/cases/pr/C96C48_hybatmDA.yaml`?

ClaraDraper-NOAA replied:
I asked @CatherineThomas-NOAA about this, and I think she's planning to include it in the GFSv17 test. Would this suffice? 

It's useful for me to have the test there, but if it's cludging up your tests I don't mind removing it.

CoryMartin-NOAA replied:
I think we just ensure that the hybatmDA test includes the soil DA work, no need to duplicate it then

ClaraDraper-NOAA replied:
I can turn on the soil DA in the hybatmDA test, and delete the current test. This will require establishing a new yaml for the combined test, since gfs_defaults_ci.yaml is used by multiple tests that don't have ensembles (so the soil DA will fail). 

What does this do:   DO_TEST_MODE: "YES"  (This is currently NO for the snowDA test, and YES for the hybatm)

CoryMartin-NOAA replied:
@DavidNew-NOAA can you give Clara an explanation of how this is used combined with GDASApp?

DavidNew-NOAA replied:
@ClaraDraper-NOAA `DO_TEST_MODE: "YES"` turns on JEDI application tests. It's referenced in the JCB algorithm YAMLs for various jobs, telling JCB what the parameters (reference file, test tolerances, etc) are for the associated JEDI application.

For atm 3dvar for example: 

https://github.com/NOAA-EMC/GDASApp/blob/91b6c1d25fabc3eb439ae66b925e0536b8f022ac/test/gw-ci/atm/jcb-prototype_3dvar_ufs_hybatmDA.yaml.j2#L36

For this example, those parameters would go in here:

https://github.com/NOAA-EMC/jcb-algorithms/blob/1c39f50c41a51e15e8e839ec86aff0cb71b7406f/3dvar.yaml.j2#L100

ClaraDraper-NOAA replied:
Thanks @DavidNew-NOAA. Is this still relevant here, even though this test is not running JEDI? 

If so, I assume I'll need to create new baselines once I add the soil DA to the test.

DavidNew-NOAA replied:
No, not really relevant here, though in one case it's used for testing in GSI, reducing the number of iterations in the minimizer:

https://github.com/NOAA-EMC/global-workflow/blob/6021dd2f12ba76bced2166bff0ba4e2be6c72018/parm/config/gfs/config.anal#L33

So you can more or less use it for whatever you want in the context of testing.

ClaraDraper-NOAA replied:
Thanks - I left DO_TEST_MODE: "YES" to keep the faster minimization.

ClaraDraper-NOAA committed:
initialize in hofx_2m_sfcfile in exgdas_enkf_update.sh

CatherineThomas-NOAA reviewed (dismissed):
We've been using this branch for the GFSv17 full res tests on WCOSS2 and it's been working well.  Save for any change to the CI tests, code changes look good to me.  Approve.

ClaraDraper-NOAA committed:
Merge branch 'develop' into feature/mv_regridding

ClaraDraper-NOAA committed:
Merged soil DA into hybatm test, and removed separate soil DA test.

ClaraDraper-NOAA committed:
coding norms.

ClaraDraper-NOAA committed:
Removed soilDA test yaml.

ClaraDraper-NOAA commented:
All, I've deleted the separate soil DA test, and added the DO_GSISOILDA option to the atmos 3DVar test. I think this is ready to go now.

ClaraDraper-NOAA commented:
> All, I've deleted the separate soil DA test, and added the DO_GSISOILDA option to the atmos 3DVar test. I think this is ready to go now.

I'll confirm once I've run the test successfuly.

ClaraDraper-NOAA commented:
Confirming that the updated C96C48_hybatmDA and the C96_atm3DVar test have run successfully.

ClaraDraper-NOAA committed:
Merge branch 'develop' into feature/mv_regridding

ClaraDraper-NOAA commented:
@CoryMartin-NOAA @CatherineThomas-NOAA Could one of you please review / approve this?

CoryMartin-NOAA reviewed (dismissed):
Looks great, thanks @ClaraDraper-NOAA

CatherineThomas-NOAA reviewed (dismissed):
Looks great.  Thanks Clara!

CatherineThomas-NOAA commented:
FYI @aerorahul @KateFriedman-NOAA @DavidHuber-NOAA:  
This PR is good to proceed.  @ClaraDraper-NOAA will be on leave for 3 weeks starting April 28, so it would be good to get any additional needed changes sorted before then. Thanks!

KateFriedman-NOAA commented:
WCOSS2 CI completed successfully, however two gempak jobs failed twice over weekend but succeeded on third attempts this morning.
```
Mon Apr 21 14:00:42 UTC 2025
******** C48_ATM_3452 ********
   CYCLE         STATE           ACTIVATED              DEACTIVATED     
202103231200        Done    Apr 18 2025 14:09:58    Apr 18 2025 15:15:33

******** C48mx500_3DVarAOWCDA_3452 ********
   CYCLE         STATE           ACTIVATED              DEACTIVATED     
202103241800        Done    Apr 18 2025 14:09:59    Apr 18 2025 14:25:41
202103250000        Done    Apr 18 2025 14:09:59    Apr 18 2025 16:00:30

******** C48mx500_hybAOWCDA_3452 ********
   CYCLE         STATE           ACTIVATED              DEACTIVATED     
202103241800        Done    Apr 18 2025 14:10:01    Apr 18 2025 14:25:45
202103250000        Done    Apr 18 2025 14:10:01    Apr 18 2025 15:35:42

******** C48_S2SW_3452 ********
   CYCLE         STATE           ACTIVATED              DEACTIVATED     
202103231200        Done    Apr 18 2025 14:10:03    Apr 18 2025 15:20:34

******** C48_S2SWA_gefs_3452 ********
   CYCLE         STATE           ACTIVATED              DEACTIVATED     
202103231200        Done    Apr 18 2025 14:10:05    Apr 18 2025 15:40:51

******** C96_atm3DVar_extended_3452 ********
   CYCLE         STATE           ACTIVATED              DEACTIVATED     
202112201800        Done    Apr 18 2025 14:10:08    Apr 18 2025 14:25:58
202112210000        Done    Apr 18 2025 14:10:08    Apr 18 2025 19:05:41
202112210600        Done    Apr 18 2025 14:10:08    Apr 18 2025 19:25:36
202112211200        Done    Apr 18 2025 14:30:59    Apr 21 2025 14:00:45
202112211800        Done    Apr 18 2025 19:10:39    Apr 18 2025 23:55:33

******** C96C48_hybatmaerosnowDA_3452 ********
   CYCLE         STATE           ACTIVATED              DEACTIVATED     
202112201200        Done    Apr 18 2025 14:10:10    Apr 18 2025 14:31:03
202112201800        Done    Apr 18 2025 14:10:10    Apr 18 2025 16:46:13
202112210000        Done    Apr 18 2025 14:10:10    Apr 18 2025 16:35:52

******** C96C48_hybatmDA_3452 ********
   CYCLE         STATE           ACTIVATED              DEACTIVATED     
202112201800        Done    Apr 18 2025 14:10:12    Apr 18 2025 14:26:06
202112210000        Done    Apr 18 2025 14:10:12    Apr 18 2025 16:40:42
202112210600        Done    Apr 18 2025 14:10:12    Apr 18 2025 16:25:46

******** C96mx100_S2S_3452 ********
   CYCLE         STATE           ACTIVATED              DEACTIVATED     
199405010000        Done    Apr 18 2025 14:10:14    Apr 18 2025 15:45:51
```

aerorahul commented on dev/jobs/sfcanl.sh:
```diff
@@ -4,10 +4,7 @@ set -x
-# until g-w issue 3390 is resolved.
-source "${HOMEgfs}/ush/load_fv3gfs_modules.sh"
-#source "${HOMEgfs}/ush/load_ufsda_modules.sh"
+. "${HOMEgfs}/ush/load_fv3gfs_modules.sh"
```
```suggestion
source "${HOMEgfs}/ush/load_fv3gfs_modules.sh"
```

aerorahul committed:
Update dev/jobs/sfcanl.sh

aerorahul reviewed (approved):
looks good.
Thanks @ClaraDraper-NOAA

aerorahul committed:
Merge branch 'develop' into feature/mv_regridding
